shardtree/src/lib.rs

```diff
@@ -0,0 +1,2851 @@
+    /// be considered to be a somewhat expensive operation.
+    pub fn marked_positions(&self, root_addr: Address) -> BTreeSet<Position> {
+        match &self.0 {
+            Node::Parent { left, right, .. } => {
+                // We should never construct parent nodes where both children are Nil.
+                // While we could handle that here, if we encountered that case it would
+                // be indicative of a programming error elsewhere and so we assert instead.
+                assert!(!(left.0.is_nil() && right.0.is_nil()));
+                let (left_root, right_root) = root_addr
+                    .children()
+                    .expect("A parent node cannot appear at level 0");
+
+                let mut left_incomplete = left.marked_positions(left_root);
+                let mut right_incomplete = right.marked_positions(right_root);
+                left_incomplete.append(&mut right_incomplete);
+                left_incomplete
+            }
+            Node::Leaf {
+                value: (_, retention),
+            } => {
+                let mut result = BTreeSet::new();
+                if root_addr.level() == 0.into() && retention.is_marked() {
+                    result.insert(Position::from(root_addr.index()));
+                }
+                result
+            }
+            Node::Nil => BTreeSet::new(),
+        }
+    }
+
+    /// Prunes the tree by hashing together ephemeral sibling nodes.
+    ///
+    /// `level` must be the level of the root of the node being pruned.
+    pub fn prune(self, level: Level) -> Self {
+        match self {
+            Tree(Node::Parent { ann, left, right }) => Tree::unite(
+                level,
+                ann,
+                left.as_ref().clone().prune(level - 1),
+                right.as_ref().clone().prune(level - 1),
+            ),
+            other => other,
+        }
+    }
+
+    /// Merge two subtrees having the same root address.
+    ///
+    /// The merge operation is checked to be strictly additive and returns an error if merging
+    /// would cause information loss or if a conflict between root hashes occurs at a node. The
+    /// returned error contains the address of the node where such a conflict occurred.
+    pub fn merge_checked(self, root_addr: Address, other: Self) -> Result<Self, Address> {
+        #[allow(clippy::type_complexity)]
+        fn go<H: Hashable + Clone + PartialEq>(
+            addr: Address,
+            t0: PrunableTree<H>,
+            t1: PrunableTree<H>,
+        ) -> Result<PrunableTree<H>, Address> {
+            // Require that any roots the we compute will not be default-filled by picking
+            // a starting valid fill point that is outside the range of leaf positions.
+            let no_default_fill = addr.position_range_end();
+            match (t0, t1) {
+                (Tree(Node::Nil), other) => Ok(other),
+                (other, Tree(Node::Nil)) => Ok(other),
```

Comment:
These can be merged together.

```suggestion
            (Tree(Node::Nil), other) | (other, Tree(Node::Nil)) => Ok(other),
```